### ISSUE TITLE
feature: produce deserialized GeoJSON from OGRGeometry in python

### ISSUE THREAD

**Issue by craigds** [OSGeo/gdal#2877]
## Current behaviour

At present `g.ExportToJson()` produces a string containing a GeoJSON geometry. 

In most places a GeoJSON geometry by itself is not very useful. If the caller is serializing an entire feature or layer, they need to deserialize the JSON and then serialize the feature/layer as a whole.:

```python
from osgeo import ogr
import json

feature = ...
g = ogr.CreateGeometryFromWkb(feature.pop('wkb'))
g_json = json.loads(g.ExportToJson())

f = {
    "type": "Feature",
    "geometry": g_json,
    "id": feature.pop('id'),
    "properties": feature
}

return json.dumps(f)
```

This seems like a waste of CPU cycles.

## Proposed behaviour

Instead, it would be useful if OGR could return the geometry as a dict (not serialized to a string), so it doesnt' need to always be deserialized by the caller and then reserialized in a wider context.

`g.ExportToJson(serialize=False)` would be a reasonable API for this.

I've only used the python bindings; I haven't looked at what other bindings do.

## Operating system

any (macos, fwiw)

## GDAL version and provenance

3.0.3

**Comment by rouault** [comment]
> return the geometry as a dict (not serialized to a string)

Hum, the issue I see is that is no such thing as a "dict" in C

**Comment by rcoup** [comment]
I guess would need to be SWIG specific native code that serialises to a native PyDictObject?

**Comment by rouault** [comment]
> I guess would need to be SWIG specific native code that serialises to a native PyDictObject?

hum, so the binding would be "intelligent" to take a OGRGeometryH and build the dict by using the OGR C API to figure out if it is a point, a linestring, etc ? That could work, but that would be a duplication of the logic in the geojson driver. I wouldn't be super enthusiastic about that.

**Comment by rcoup** [comment]
I think I'd prefer the approach of an internal serialisation callback in the GeoJSON driver/JSON utilities into the Python bindings to assemble `Py*` objects if `serialize==false`. So lower-level than an OGRGeometryH.

Is libjson-c used for serializing as well? (ie: is there switching infrastructure already between internal & json-c?)

**Comment by sgillies** [comment]
Isn't this already possible using the OGR Python API? Like

```
from osgeo import ogr

g = ogr.CreateGeometryFromWkb(feature.pop('wkb'))
f = ogr.Feature(...)  # using your layer definition
f.SetGeometryDirectly(g)
return f.ExportToJSON()
```

If you're trying to avoid the OGR Feature class, you could try using geomet to convert the geometry from wkb to a Python GeoJSON-like dict.

```
from geomet import wkb
import json

feature = ...
geom = wkb.loads(feature.pop('wkb'))

f = {
    "type": "Feature",
    "geometry": geom,
    "id": feature.pop('id'),
    "properties": feature
}

return json.dumps(f)
```

I suggest this because I'm concerned about the complexity we'd be adding to OGR's GeoJSON exporter. Also, I feel like changing the type of output of C API function depending on a flag makes the API more confusing.

**Comment by rcoup** [comment]
> Isn't this already possible using the OGR Python API? Like

Yes, `Feature.ExportToJson(as_object=True)`, but [underneath](https://github.com/OSGeo/gdal/blob/master/gdal/swig/python/osgeo/ogr.py#L4573-L4574) it calls `Geometry.ExportToJson()` then `json.loads()`

> If you're trying to avoid the OGR Feature class, you could try using geomet to convert the geometry from wkb to a Python GeoJSON-like dict.

Not so much that as avoiding encoding+decoding+encoding when incorporating OGR geometries as GeoJSON into larger object structures like API responses.

> Also, I feel like changing the type of output of C API function depending on a flag makes the API more confusing.

I agree, I'm not particularly excited about overloading the same method with different return types either.

> I suggest this because I'm concerned about the complexity we'd be adding to OGR's GeoJSON exporter. 

So, I had a read through `ogrgeojsonwriter.cpp`: it is tightly coupled to json-c, but since json-c builds an in-memory tree of objects, this all becomes _massively_ simpler:

* the existing `OGRGeoJSONWriteGeometry()`/`OGRGeoJSONWriteFeature()` functions do _all_ the work of assembling the GeoJSON structures into a `json_object`, and they wouldn't need changing at all.
* some new code that iterates over a `json_object` to assemble CPython objects (lets say `json_object_to_py_object()`). This is generic and could live in the Python bindings.
* a method similar to `OGR_G_ExportToJsonEx()` that calls `OGRGeoJSONWriteGeometry()` then the new `json_object_to_py_object()` in place of `json_object_to_json_string()`. This can live in the Python bindings too.
* additional methods that follow the same pattern for Features or whatever else
